### DISCUSSION TITLE
Namespaces

### DISCUSSION THREAD

**Discussion by aganm** [raysan5/raylib#2057] [Ideas]
One thing that bugs me a bit about raylib's API is that it doesn't use namespaces. By namespace I mean a prefix in front of each function. For example, SDL uses the `SDL_` prefix:
```
SDL_CreateWindow()
SDL_CreateWindowAndRenderer()
SDL_CreateWindowFrom()
```
Opengl uses `gl` prefix
```
glAccum()
glAddSwapHintRect()
glAlphaFunc()
```

The issue that it causes me is that have I have no way to get a list of all the raylib functions from the auto completion feature of my editor. For example:
```
InitWindow()
WindowShouldClose();                                      
CloseWindow();   
```
None of these share a common namespace so I cannot make use of the autocompletion to 'discover' the functions, I have to know in advance what I'm gonna type.

Having a `Ray_` namespace could be an obvious choice. For example:
```
void Ray_InitWindow(int width, int height, const char *title);          
void Ray_DrawPixel(int posX, int posY, Color color);             
void Ray_UploadMesh(Mesh *mesh, bool dynamic);  
```

Or Opengl style with a `rl` namespace
```
void rlInitWindow(int width, int height, const char *title);          
void rlDrawPixel(int posX, int posY, Color color);             
void rlUploadMesh(Mesh *mesh, bool dynamic);  
```

But, since raylib's API is clearly separated into different modules, each module could have its own namespace. All core functions could start with `R_` for raylib core, all texture functions could start with `T_`, for texture, all text functions could start with 'F_' for font, etc. For example:
```
    // raylib core
    void R_InitWindow(int width, int height, const char *title);          
    bool R_WindowShouldClose(void);
    void R_CloseWindow(void);

    // shapes
    void S_DrawPixel(int posX, int posY, Color color);                                               
    void S_DrawPixelV(Vector2 position, Color color);                                                  
    void S_DrawLine(int startPosX, int startPosY, int endPosX, int endPosY, Color color); 

   // textures
    Image T_ImageCopy(Image image);                                      
    Image T_ImageFromImage(Image image, Rectangle rec);        
    Image T_ImageText(const char *text, int fontSize, Color color);   

    // models
    void M_UploadMesh(Mesh *mesh, bool dynamic);  
    void M_UpdateMeshBuffer(Mesh mesh, int index, void *data, int dataSize, int offset);  
    void M_DrawMesh(Mesh mesh, Material material, Matrix transform);                      
```

I see two upsides to having this. It reduces if not removes the risk of name collision with system libraries. And it makes autocompletion useful again because I can just type `R_` and get a list of all the raylib core functions. 

What do you think? Also I'd be interested to hear about why raylib was designed without a namespace, is there any advantage to it not having a namespace?

**Comment by raysan5** [comment]
@aganm Actually this is a topic that has been disussed/commented several times in raylib Discord forums.

Avoiding prefixes was a personal decision; raylib was highly inspired by WinBGI and XNA and I tried to follow similar namings, it was simpler and clearer for my students. Afaik there was no other game library taking that approach.

I explored adding prefixes and following other naming conventions in others libraries I created. Actually, `rlgl` library used by raylib internally (and being one of the main modules of raylib) uses the `rl*()` prefix. 

At this point adding prefixes to raylib would require breaking compatibility with 8 years of library existence and many codebases. I'm afraid I have no plans for that.

**Comment by orcmid** [comment]
@aganm We're talking about a notional namespace and not a technological one.  So it is a convention, like "I" on the beginning of Interface names.  (Bites his tongue about Hungarian notation and similar efforts.)

I would think that VS Code might be helpful, although it needs to have the raylib headers set in an environment variable or in a workspace setting before Intellisense can create completions and also offer choices.  Likewise in recent Visual Studio releases, for that matter.

At the moment, I get squiggles on `"raylib.h"` and `<raylib.h>` until I find a way for the command-line environment and C/C++ Intellisense to access  `raylib.h` and others that might be referenced.

I think it is useful to print out the "cheat sheet" too.  I suppose there needs to be a v4 of that and life is going to be messy for a while as breaking changes pile up and users stay on v3.7 for stability until maybe v4.1 :).

**Reply by orcmid** [reply]
I wonder what the available C++ wrappers do in this respect, since actual (technical) namespaces can be introduced.
